site/src/pages/UserSettingsPage/NotificationsPage/NotificationsPage.tsx

```diff
@@ -0,0 +1,238 @@
+};
+
+const PreferenceSwitch: FC<PreferenceSwitchProps> = ({
+  id,
+  disabled,
+  onToggle,
+}) => {
+  const { user } = useAuthenticated();
+  const queryClient = useQueryClient();
+  const updatePreferences = useMutation(
+    updateUserNotificationPreferences(user.id, queryClient),
+  );
+
+  return (
+    <Switch
+      id={id}
+      size="small"
+      checked={!disabled}
+      onChange={async (_, checked) => {
+        await updatePreferences.mutateAsync({
+          template_disabled_map: onToggle(checked),
+        });
+        displaySuccess("Notification preferences updated");
+      }}
+    />
+  );
+};
+
+export const NotificationsPage: FC = () => {
+  const { user, permissions } = useAuthenticated();
+  const [disabledPreferences, templatesByGroup, dispatchMethods] = useQueries({
+    queries: [
+      {
+        ...userNotificationPreferences(user.id),
+        select: selectDisabledPreferences,
+      },
+      {
+        ...systemNotificationTemplates(),
+        select: (data: NotificationTemplate[]) => {
+          const groups = selectTemplatesByGroup(data);
+          return permissions.viewDeploymentValues
+            ? groups
+            : {
+                // Members only have access to the "Workspace Notifications" group
+                ["Workspace Events"]: groups["Workspace Events"],
+              };
+        },
+      },
+      notificationDispatchMethods(),
+    ],
+  });
+  const ready =
+    disabledPreferences.data && templatesByGroup.data && dispatchMethods.data;
+
+  return (
+    <>
+      <Helmet>
+        <title>{pageTitle("Notifications Settings")}</title>
+      </Helmet>
+      <Section
+        title="Notifications"
+        description="Configure your notification preferences. Icons on the right of each notification indicate delivery method, either SMTP or Webhook."
+        layout="fluid"
+      >
+        {ready ? (
+          <Stack spacing={3}>
+            {Object.entries(templatesByGroup.data).map(([group, templates]) => {
+              const allDisabled = templates.some((tpl) => {
+                return disabledPreferences.data[tpl.id] === true;
+              });
+
+              return (
+                <Card
+                  variant="outlined"
+                  css={{ background: "transparent" }}
+                  key={group}
+                >
+                  <List>
+                    <ListItem css={styles.listHeader}>
+                      <ListItemIcon>
+                        <PreferenceSwitch
+                          id={group}
+                          disabled={allDisabled}
+                          onToggle={(checked) => {
+                            const updated = { ...disabledPreferences.data };
+                            for (const tpl of templates) {
+                              updated[tpl.id] = !checked;
+                            }
+                            return updated;
+                          }}
```

Comment:
To be clear, `onChange` and all the event handlers *receive* an event as input, but they don't return anything. This is the type React has:
```tsx
type React.EventHandler<E extends SyntheticEvent<any>> = (event: E) => void;
```
In real HTML, you can have the event handlers return booleans to affect event bubbling behavior, but React shaves that off for the synthetic events.

Like, if I have a component like this:
```tsx
function MyComponent () {
  const [color, setColor] = useState("blue");
  return <button onClick={() => setMyName("red"}>Click me</button>;
}
```
I don't think anyone would expect the button to have any way to affect `MyComponent`'s state, beyond the `onClick`. But even then, `MyComponent` is in full control of the event handler definition – it fully prescribes how its own state is allowed to change. `MyComponent` has a dependency on `button`, but everything related to its state is self-contained in the one component. `button` doesn't have direct access to the state setter, so all it can do is follow the very restricted set of instructions it's given

But with the switches, the main state is being split across four different places: the RQ cache living outside the UI tree, the parent page component, the individual switches, and the mutation key factory function. It felt like I had to jump through a lot of hoops (and two different files) to figure out exactly how the state was changing 

I don't think there's a way to fully consolidate the state in one place, just because we do need RQ, but four places for one main piece of state feels like a bit much to me

Reply:
But as far as conventions around one-way data-binding/unidirectional data flow, I tried looking on React.dev, and I'm surprised there wasn't too much dedicated information on it. There's [Thinking in React](https://react.dev/learn/thinking-in-react), which is always a classic, but I think it's the sort of thing that's never explicitly shouted out, even though it informs pretty much every lesson on the site

The main reason I know about all this is because I just happen to know that React was designed around the flux architecture pattern

Reply:
I think I better understand your concerns now. Thanks for explaining them to me. I will fix it.

Reply:
Done!